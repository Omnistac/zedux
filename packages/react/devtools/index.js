if (process.env.NODE_ENV !== 'development') {
  module.exports = {
    getStateHub: function getStateHub() {},
    StateHub: function StateHub() {
      return null
    },
  }
} else {
  module.exports = require('./dev')
}
