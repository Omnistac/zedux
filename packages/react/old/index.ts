export * from './AppProvider'
export { globalStore } from './global/store'
export * from './createAtom'
export * from './createMolecule'
export * from './types'
export * from './useAtom'
export * from './useMolecule'
export * from './useStore'
