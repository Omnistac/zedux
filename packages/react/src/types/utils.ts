import { AtomTemplateBase } from '../classes/templates/AtomTemplateBase'
import { AtomInstance } from '../classes/instances/AtomInstance'
import { AtomInstanceBase } from '../classes/instances/AtomInstanceBase'

export type AnyAtomInstance = AtomInstance<any>
export type AnyAtomInstanceBase = AtomInstanceBase<
  any,
  AtomTemplateBase<any, AtomInstance<any>>
>
export type AnyAtomTemplate = AtomTemplateBase<any, any>
