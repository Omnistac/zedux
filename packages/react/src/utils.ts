import { AnyAtomTemplate, Ecosystem } from '@zedux/atoms'
import React, { createContext } from 'react'

export const ecosystemContext = createContext('@@global')

/**
 * These are copied from packages/atoms/src/utils/general.ts
 *
 * IMPORTANT: keep these in-sync with the ones in the atoms package
 */
export const Explicit = 1
export const External = 2
export const Static = 4
// export const Deferred = 8

export const destroyed = Symbol.for(`@@zedux/destroyed`)

export const getReactContext = (
  ecosystem: Ecosystem,
  atom: AnyAtomTemplate
) => {
  const reactStorage: {
    contexts: WeakMap<typeof createContext, Record<string, React.Context<any>>>
  } = (ecosystem._storage.react ||= {})

  const contexts = (reactStorage.contexts ||= new WeakMap())
  const windowContexts =
    contexts.get(createContext) ||
    (contexts.set(createContext, {}).get(createContext) as Record<
      string,
      React.Context<any>
    >)

  return (windowContexts[atom.key] ||= createContext(
    undefined
  ) as React.Context<any>)
}
