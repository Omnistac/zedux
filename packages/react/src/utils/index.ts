export * from './general'
export * from './types'
