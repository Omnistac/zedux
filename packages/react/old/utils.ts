import { Atom, Scope } from './types'

let idCounter = 0

const generateId = () => {
  idCounter += 1
  return idCounter
}

const stringifyParam = (param: any): string => {
  if (Array.isArray(param)) {
    return `[${param.map(child => stringifyParam(child)).join(',')}]`
  }

  if (param && typeof param === 'object') {
    return `{${Object.entries(param)
      .map(([key, val]) => `${key}: ${stringifyParam(val)}`)
      .sort((a, b) => a.localeCompare(b))
      .join(',')}}`
  }

  return JSON.stringify(param)
}

export const generateAppId = () => `app-${generateId()}`
export const generateImplementationId = () => `im-${generateId()}`
export const generateInstanceId = () => `in-${generateId()}`
export const generateLocalId = () => `lo-${generateId()}`

export const getFullKey = (atom: Atom, params?: any[]) => {
  if (atom.scope === Scope.local) return generateLocalId()

  if (!params?.length) return atom.key

  return `${atom.key}---${stringifyParam(params)}`
}
