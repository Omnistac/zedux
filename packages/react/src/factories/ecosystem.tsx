import { Ecosystem } from '../classes/Ecosystem'
import { EcosystemConfig } from '../types'

export const ecosystem = (config: EcosystemConfig) => new Ecosystem(config)
