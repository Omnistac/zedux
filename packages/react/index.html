<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Zedux Dev</title>
  </head>
  <body>
    <div id="root"></div>
    <script type="module" src="/src/index.ts"></script>
    <script type="module">
      import { atom, useAtomState } from './src/index'
      import { createElement } from 'react'
      import { render } from 'react-dom'

      const greetingAtom = atom('greeting', 'Hello, World!')

      const App = () => {
        const [state, setState] = useAtomState(greetingAtom)

        return createElement('div', null,
          createElement('p', null, state),
          createElement('input', {
            onChange: event => setState(event.currentTarget.value),
            value: state
          })
        )
      }

      render(createElement(App), document.getElementById('root'))
    </script>
  </body>
</html>
