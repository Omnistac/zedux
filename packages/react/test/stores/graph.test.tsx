import { fireEvent } from '@testing-library/dom'
import { act } from '@testing-library/react'
import {
  injectAtomGetters,
  injectAtomValue,
  useAtomInstance,
  useAtomValue,
  injectAtomInstance,
  injectRef,
  AtomGetters,
  EvaluationReason,
} from '@zedux/react'
import {
  api,
  atom,
  AtomInstanceType,
  createStore,
  injectStore,
  ion,
} from '@zedux/stores'
import React from 'react'
import { ecosystem, getNodes, snapshotNodes } from '../utils/ecosystem'
import { renderInEcosystem } from '../utils/renderInEcosystem'
import { Eventless } from '@zedux/react/utils'

const atom1 = atom('atom1', () => 1)
const atom2 = atom('atom2', () => 2)
const atom3 = atom('atom3', () => 3)

const atom4 = atom('atom4', () => {
  const switchStore = injectStore(true)
  const sumStore = injectStore(0)
  const store = injectStore(() =>
    createStore({ switch: switchStore, sum: sumStore })
  )
  const { get } = injectAtomGetters()

  const one = get(atom1)
  const two = switchStore.getState() ? get(atom2) : get(atom3)

  // won't cause a reevaluation loop
  sumStore.setState(one + two)

  return api(store).setExports({
    toggle: () => switchStore.setState(val => !val),
  })
})

describe('graph', () => {
  test('injectAtomGetters', async () => {
    jest.useFakeTimers()

    function Test() {
      const { sum } = useAtomValue(atom4)
      const { toggle } = useAtomInstance(atom4).exports

      return (
        <>
          <div data-testid="sum">{sum}</div>
          <button onClick={toggle}>toggle</button>
        </>
      )
    }

    const { findByTestId, findByText } = renderInEcosystem(<Test />)

    const div = await findByTestId('sum')
    expect(div).toHaveTextContent('3')

    const expectedEdges = {
      flags: Eventless,
      operation: 'get',
    }

    const nodes = getNodes()

    expect(nodes.atom1).toEqual(
      expect.objectContaining({
        observers: {
          atom4: expectedEdges,
        },
      })
    )

    expect(nodes.atom2).toEqual(
      expect.objectContaining({
        observers: {
          atom4: expectedEdges,
        },
      })
    )

    expect(nodes.atom3).toBeUndefined()

    expect(
      [...ecosystem.n.get('atom4')!.s.keys()].map(node => node.id)
    ).toEqual(['atom1', 'atom2'])

    const button = await findByText('toggle')

    act(() => {
      fireEvent.click(button)
      jest.runAllTimers()
    })

    const nodes2 = getNodes()

    expect(div).toHaveTextContent('4')

    expect(nodes2.atom1).toEqual(
      expect.objectContaining({
        observers: {
          atom4: expectedEdges,
        },
      })
    )

    expect(nodes2.atom2).toEqual(
      expect.objectContaining({
        observers: {},
      })
    )

    expect(nodes2.atom3).toEqual(
      expect.objectContaining({
        observers: {
          atom4: expectedEdges,
        },
      })
    )

    expect(nodes2.atom4).toEqual(
      expect.objectContaining({
        sources: {
          atom1: expect.any(Object),
          atom3: expect.any(Object),
        },
      })
    )

    expect(ecosystem.viewGraph()).toMatchSnapshot()
    expect(ecosystem.viewGraph('bottom-up')).toMatchSnapshot()
    expect(ecosystem.viewGraph('top-down')).toMatchSnapshot()
  })

  test('getInstance(atom) returns the instance', () => {
    const evaluations: number[] = []

    const ion1 = ion('ion1', ({ getInstance }) => {
      const instance1 = getInstance(atom1)
      const instance2 = getInstance(atom2)

      evaluations.push(instance1.store.getState())

      return api(
        instance1.store.getState() + instance2.store.getState()
      ).setExports({
        set: (val: number) => getInstance(atom1).setState(val + 1),
      })
    })

    const ionInstance = ecosystem.getInstance(ion1)

    expect(ecosystem.findAll().map(({ id }) => id)).toEqual([
      'atom1',
      'atom2',
      'ion1',
    ])

    snapshotNodes()
    expect(evaluations).toEqual([1])

    ionInstance.exports.set(11)

    expect(evaluations).toEqual([1])
    expect(ionInstance.store.getState()).toBe(3)
    expect(ecosystem.getInstance(atom1).store.getState()).toBe(12)
  })

  test('on reevaluation, get() updates the graph', () => {
    jest.useFakeTimers()
    let useB = true
    const atomA = atom('a', () => 'a')
    const atomB = atom('b', (param: string) => param)
    const atomC = atom('c', 'c')
    const atomD = ion('d', ({ get }) => {
      const a = injectAtomValue(atomA)
      const b = useB ? get(atomB, ['b']) : get(atomC)

      return a + b
    })

    const instance = ecosystem.getInstance(atomD)

    snapshotNodes()

    useB = false
    instance.invalidate()
    jest.runAllTimers()

    snapshotNodes()
  })

  test('atom instances can be passed as atom params', () => {
    const instance1 = ecosystem.getInstance(atom1)

    const testAtom = atom(
      'test',
      (instance: AtomInstanceType<typeof atom1>) => {
        const subRef = injectRef(false)
        const asDep = injectAtomInstance(instance) // register static dep
        const { get } = injectAtomGetters()

        if (subRef.current) get(asDep)

        return api(asDep.getState()).setExports({ subRef })
      }
    )

    const instance = ecosystem.getInstance(testAtom, [instance1])

    expect(instance.getState()).toBe(1)

    instance1.setState(2)

    expect(instance.getState()).toBe(1)

    instance.exports.subRef.current = true
    instance.invalidate()

    expect(instance.getState()).toBe(2)

    instance1.setState(3)

    expect(instance.getState()).toBe(3)
  })

  test('simultaneously-scheduled selector evaluations cause one evaluation', () => {
    const atom1 = atom('1', 'a')
    const atom2 = ion('2', ({ get }) => get(atom1) + 'b')

    let why: EvaluationReason[] | undefined

    const selector1 = ({ get }: AtomGetters) => {
      why = ecosystem.why()
      return get(atom1) + get(atom2)
    }

    const selectorInstance = ecosystem.getNode(selector1)
    const atomInstance = ecosystem.getInstance(atom1)

    expect(selectorInstance.v).toBe('aab')
    expect(why).toHaveLength(0)

    atomInstance.setState('aa')

    expect(selectorInstance.v).toBe('aaaab')
    expect(why).toHaveLength(2)
    expect(why).toEqual([
      {
        newState: 'aa',
        oldState: 'a',
        operation: 'get',
        reasons: undefined,
        source: expect.any(Object),
        type: 'change',
      },
      {
        newState: 'aab',
        oldState: 'ab',
        operation: 'get',
        reasons: [
          {
            newState: 'aa',
            oldState: 'a',
            operation: 'get',
            reasons: undefined,
            source: expect.any(Object),
            type: 'change',
          },
        ],
        source: expect.any(Object),
        type: 'change',
      },
    ])
  })
})
