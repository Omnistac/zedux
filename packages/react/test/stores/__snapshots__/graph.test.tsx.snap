// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graph getInstance(atom) returns the instance 1`] = `
{
  "atom1": {
    "className": "AtomInstance",
    "observers": {
      "ion1": {
        "flags": 4,
        "operation": "getNode",
        "p": undefined,
      },
    },
    "sources": {},
    "state": 1,
    "status": "Active",
    "weight": 1,
  },
  "atom2": {
    "className": "AtomInstance",
    "observers": {
      "ion1": {
        "flags": 4,
        "operation": "getNode",
        "p": undefined,
      },
    },
    "sources": {},
    "state": 2,
    "status": "Active",
    "weight": 1,
  },
  "ion1": {
    "className": "AtomInstance",
    "observers": {},
    "sources": {
      "atom1": {
        "flags": 4,
        "operation": "getNode",
        "p": undefined,
      },
      "atom2": {
        "flags": 4,
        "operation": "getNode",
        "p": undefined,
      },
    },
    "state": 3,
    "status": "Active",
    "weight": 3,
  },
}
`;

exports[`graph injectAtomGetters 1`] = `
{
  "Test-:r0:": {
    "dependencies": [
      {
        "key": "atom4",
        "operation": "useAtomValue",
      },
    ],
    "dependents": [],
    "weight": 3,
  },
  "Test-:r1:": {
    "dependencies": [
      {
        "key": "atom4",
        "operation": "useAtomInstance",
      },
    ],
    "dependents": [],
    "weight": 0,
  },
  "atom1": {
    "dependencies": [],
    "dependents": [
      {
        "key": "atom4",
        "operation": "get",
      },
    ],
    "weight": 1,
  },
  "atom2": {
    "dependencies": [],
    "dependents": [],
    "weight": 1,
  },
  "atom3": {
    "dependencies": [],
    "dependents": [
      {
        "key": "atom4",
        "operation": "get",
      },
    ],
    "weight": 1,
  },
  "atom4": {
    "dependencies": [
      {
        "key": "atom1",
        "operation": "get",
      },
      {
        "key": "atom3",
        "operation": "get",
      },
    ],
    "dependents": [
      {
        "key": "Test-:r0:",
        "operation": "useAtomValue",
      },
      {
        "key": "Test-:r1:",
        "operation": "useAtomInstance",
      },
    ],
    "weight": 3,
  },
}
`;

exports[`graph injectAtomGetters 2`] = `
{
  "Test-:r0:": {
    "atom4": {
      "atom1": {},
      "atom3": {},
    },
  },
  "Test-:r1:": {
    "atom4": {
      "atom1": {},
      "atom3": {},
    },
  },
  "atom2": {},
  "atom4": {
    "atom1": {},
    "atom3": {},
  },
}
`;

exports[`graph injectAtomGetters 3`] = `
{
  "atom1": {
    "atom4": {
      "Test-:r0:": {},
      "Test-:r1:": {},
    },
  },
  "atom2": {},
  "atom3": {
    "atom4": {
      "Test-:r0:": {},
      "Test-:r1:": {},
    },
  },
}
`;

exports[`graph on reevaluation, get() updates the graph 1`] = `
{
  "a": {
    "className": "AtomInstance",
    "observers": {
      "d": {
        "flags": 0,
        "operation": "injectAtomValue",
        "p": undefined,
      },
    },
    "sources": {},
    "state": "a",
    "status": "Active",
    "weight": 1,
  },
  "b-["b"]": {
    "className": "AtomInstance",
    "observers": {
      "d": {
        "flags": 0,
        "operation": "get",
        "p": undefined,
      },
    },
    "sources": {},
    "state": "b",
    "status": "Active",
    "weight": 1,
  },
  "d": {
    "className": "AtomInstance",
    "observers": {},
    "sources": {
      "a": {
        "flags": 0,
        "operation": "injectAtomValue",
        "p": undefined,
      },
      "b-["b"]": {
        "flags": 0,
        "operation": "get",
        "p": undefined,
      },
    },
    "state": "ab",
    "status": "Active",
    "weight": 3,
  },
}
`;

exports[`graph on reevaluation, get() updates the graph 2`] = `
{
  "a": {
    "className": "AtomInstance",
    "observers": {
      "d": {
        "flags": 0,
        "operation": "injectAtomValue",
        "p": undefined,
      },
    },
    "sources": {},
    "state": "a",
    "status": "Active",
    "weight": 1,
  },
  "b-["b"]": {
    "className": "AtomInstance",
    "observers": {},
    "sources": {},
    "state": "b",
    "status": "Stale",
    "weight": 1,
  },
  "c": {
    "className": "AtomInstance",
    "observers": {
      "d": {
        "flags": 0,
        "operation": "get",
        "p": undefined,
      },
    },
    "sources": {},
    "state": "c",
    "status": "Active",
    "weight": 1,
  },
  "d": {
    "className": "AtomInstance",
    "observers": {},
    "sources": {
      "a": {
        "flags": 0,
        "operation": "injectAtomValue",
        "p": undefined,
      },
      "c": {
        "flags": 0,
        "operation": "get",
        "p": undefined,
      },
    },
    "state": "ac",
    "status": "Active",
    "weight": 3,
  },
}
`;
